src/mlpack/methods/reinforcement_learning/ppo_impl.hpp

```diff
@@ -0,0 +1,288 @@
+  #if ENS_VERSION_MAJOR >= 2
+  actorUpdatePolicy(NULL),
+  #endif
+  criticUpdater(std::move(updater)),
+  #if ENS_VERSION_MAJOR >= 2
+  criticUpdatePolicy(NULL),
+  #endif
+  environment(std::move(environment)),
+  totalSteps(0),
+  deterministic(false)
+{
+
+  oldActorNetwork = actorNetwork;
+  // Reset all the networks.
+  // Note: the actor and critic networks have an if condition before reset.
+  // This is because we don't want to reset a loaded(possibly pretrained) model
+  // passed using this constructor.
+  const size_t envSampleSize = environment.InitialSample().Encode().n_elem;
+  if (actorNetwork.Parameters().n_elem != envSampleSize)
+    actorNetwork.Reset(envSampleSize);
+
+  if (criticNetwork.Parameters().n_elem != envSampleSize)
+    criticNetwork.Reset(envSampleSize);
+
+  #if ENS_VERSION_MAJOR == 1
+  this->criticUpdater.Initialize(criticNetwork.Parameters().n_rows,
+                                 criticNetwork.Parameters().n_cols);
+  #else
+  this->criticUpdatePolicy = new typename UpdaterType::template
+  Policy<arma::mat, arma::mat>(this->criticUpdater,
+                               criticNetwork.Parameters().n_rows,
+                               criticNetwork.Parameters().n_cols);
+  #endif
+
+  #if ENS_VERSION_MAJOR == 1
+  this->actorUpdater.Initialize(actorNetwork.Parameters().n_rows,
+                                actorNetwork.Parameters().n_cols);
+  #else
+  this->actorUpdatePolicy = new typename UpdaterType::template
+  Policy<arma::mat, arma::mat>(this->actorUpdater,
+                               actorNetwork.Parameters().n_rows,
+                               actorNetwork.Parameters().n_cols);
+  #endif
+
+  // Copy over the actor parameters to its older self.
+  oldActorNetwork.Parameters() = actorNetwork.Parameters();
+}
+
+template<
+  typename EnvironmentType,
+  typename ActorNetworkType,
+  typename CriticNetworkType,
+  typename UpdaterType
+>
+PPO<
+  EnvironmentType,
+  ActorNetworkType,
+  CriticNetworkType,
+  UpdaterType
+>::~PPO()
+{
+  #if ENS_VERSION_MAJOR >= 2
+  delete actorUpdatePolicy;
+  delete criticUpdatePolicy;
+  #endif
+}
+
+template<
+  typename EnvironmentType,
+  typename ActorNetworkType,
+  typename CriticNetworkType,
+  typename UpdaterType
+>
+void PPO<
+  EnvironmentType,
+  ActorNetworkType,
+  CriticNetworkType,
+  UpdaterType
+>::SelectAction()
+{
+  // Get action logits for each action at current state.
+  arma::colvec actionLogit;
+  actorNetwork.Predict(state.Encode(), actionLogit);
+
+  arma::colvec prob;
+  ann::Softmax softmax;
+  softmax.Forward(actionLogit, prob);
+
+  // Calculating cumulative probablity.
+  for (size_t i = 1; i < prob.n_rows; i++)
+      prob[i] += prob[i-1];
+
+  // Sampling action from cumulative probablity.
+  double randValue = arma::randu<double>();
+  for (size_t i = 0; i<prob.n_rows; i++)
+  {
+    if (randValue <= prob[i])
+    {
+      action.action = static_cast<decltype(action.action)>(i);
+      return;
+    }
+  }
```

Comment:
Few things, normally we expect variables names to be explanatory, unless it's blatantly obvious. For L133, it's fine, but L138 not so much. 

For L128, you could just rename `prob` to `cumulativeProb`, so it's intent is clear. That way you can eliminate L132 entirely.

L138, `size_t actionIdx = 0; actionIdx < cumulativeProb.n_rows; ++actionIdx`. Also no need to declare `randValue` directly compare L140 `if (arma::randu<double> <= cumulativeProb[actionIdx])`

Reply:
Also a random comment, I happened to see this while scrolling through my emails (nice work with the PR by the way!), it would be good for consistency to use `mlpack::math::Random()` instead of `arma::randu<>()`. :+1:  But that is a super minor comment. :)

Reply:
Sure ! i will make those changes.